walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -31,22 +31,39 @@
       return spots;
     }
 
-    /** Deletes irrelevant events. */
-    eventsList.removeIf(event -> Collections.disjoint(request.getAttendees(), 
-                                                      event.getAttendees()));
+    eventsList.removeIf(event -> 
+        (Collections.disjoint(request.getAttendees(), event.getAttendees()) &&
+         Collections.disjoint(request.getOptionalAttendees(), event.getAttendees())));
 
-    if (eventsList.isEmpty() || request.getAttendees().isEmpty()) {
+    if (eventsList.isEmpty()) {
       return Arrays.asList(TimeRange.WHOLE_DAY);
     }
-    
+
     /** Sorts the events list by start time. */
     Collections.sort(eventsList, new Comparator<Event>(){
       public int compare(Event a, Event b) {
         return TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen());
       }
     });
     
+    if (request.getAttendees().isEmpty() || request.getOptionalAttendees().isEmpty()) {
+      return (Collection<TimeRange>)findMeetingSpots(eventsList, meetingDuration);
+    }
```

Comment:
this early break doesn't seem to be needed and it seems to reduce readability.
please consider removing.

Reply:
This early break is for the case where there are only optional attendees with no gaps in their schedules (test #5 above).

Reply:
When I ran the tests on Cloud Shell, test #5 would fail if this section was removed. The method would return TimeRange.WHOLE_DAY instead of an empty list. I'll add the comments above to make it more readable.

Reply:
I guess that's because you should remove disjoint optional attendees in the removeIf below?

Reply:
I think when disjoint optional attendees are removed, since there is no mandatory attendee, the method would conclude that the whole day is available. That's probably where things will go wrong.

Reply:
let me be clear.
I think you should write
```
eventsList.removeIf(event -> Collections.disjoint(
    request.getOptionalAttendees(), event.getAttendees()));
```
which remove events that are irrelevant to optional attendees.

You implementation remove events that are irrelevant to required attendees.
In test 5th, you remove all events since all events are irrelevant to the required attendees, that's why you return the whole day.

And yes, you could keep the early return, which save you an extra findMeetingQuery call, but you should add comments. And a better option is to add if together with the empty check.
```
if (spots.isEmpty() && !request.getOptionalAttendees().empty()) {
}
```

Reply:
It seems that you didn't fix this issue?

Reply:
I didn't quite get the better option above. If no spots is available for both mandatory and optional attendees, according to the requirement, the method should find spots that at least fits all mandatory attendees. Only when there is no mandatory attendees would we remove events that are irrelevant to only optional attendees, which this section is written to check. I have added comments above this section to explain its purpose.

Reply:
L57-58 should be
```
eventsList.removeIf(event -> Collections.disjoint(
    request.getOptionalAttendees(), event.getAttendees()));
```
rather than
```
eventsList.removeIf(event -> Collections.disjoint(
    request.getAttendees(), event.getAttendees()));
```
right?

Reply:
Probably not. With
`eventsList.removeIf(event -> Collections.disjoint(
    request.getAttendees(), event.getAttendees()));`
Only events already on mandatory attendees' calendars are kept. Then I could look into gaps between these already existing events and find spots to throw the requested meeting. In the 2nd version, only the schedules of optional attendees are considered.

Reply:
yes, you're right. I miss the removeIf ...